### ISSUE TITLE
DSMR: no gas usage for DSMR2

### ISSUE THREAD

**Issue by peterforeman** [esphome/issues#2517]
### The problem

Earlier the DSMR component was updated for usage with the DSMR2 standard. These changes are merged in the dev branch. I tried updating my ESP32 P1 reader with the latest ESPhome dev and now it reports the correct data - except the gas usage which is stays "unknown".

Closed DSMR2 issue: https://github.com/esphome/issues/issues/2363
PR that did not fix this problem: https://github.com/esphome/esphome/pull/2157

### Which version of ESPHome has the issue?

2021.10.0-dev

### What type of installation are you using?

Docker

### Which version of Home Assistant has the issue?

Not relevant

### What platform are you using?

ESP32

### Board

ESP32 wrover

### Component causing the issue

dsmr

### Example YAML snippet

```yaml
#######################################################################
# Use ESP32 Serial2
uart:
  tx_pin: 17
  rx_pin: 16
  baud_rate: 9600
  data_bits: 7
  parity: NONE
  stop_bits: 1

# DSMR2 has no CRC
dsmr:
  crc_check: false

sensor:
  - platform: dsmr
    gas_delivered:
      name: "Gas Delivery Total"
      id: gas_delivery_total
      state_class: "total_increasing"
      unit_of_measurement: "m³"
```

### Anything in the logs that might be useful for us?

```txt
No error messages, just no update for this specific sensor. Same for DSMR version sensor.

Text that DSMR2 produces:
0-2:24.3.0(180822110000)(08)(60)(1)(0-2:24.2.0)(m3)
(04700.692)

Note that the actual sensor data is on the next line.
```

**Comment by mmakaay** [comment]
The field in the parser code that most closely matches these values is:
```c++
DEFINE_FIELD(gas_delivered_text, String, ObisId(0, GAS_MBUS_ID, 24, 3, 0), RawField);
```
This field isn't defined as a m3 unit but as a raw field. For this to work correctly, I think that it should be a m3 field instead. But there's more. I'm not even sure that the field is recognized by the parser code currently.

The `GAS_MBUS_ID` in the parser is hard-coded to the value 1.
That might be an issue here. Your MBUS ID is 2. In the parser code that is related to measurements for "water".
Does your meter by any chance have more measurements on board than those for electricity and gas?

From the DSMR v4 specs:
```
The OBIS reference Channel number (n) will be defined by the installing order
of the M-Bus devices. So a Gas meter can be installed on channel 1, 2, 3 or 4.
```

Based on that, I would say that the parser shouldn't be looking at the MBUS ID, or at least it should expect multiple values.

**TL;DR**

* The field definition ought to be a m3 unit integer field to make this work.
* The MBUS ID should not be limited to the value of 1.

**Comment by dhover** [comment]
I don't if it helps but this is an actual telegram:

`
/KMP5 ZABF001770315713

0-0:96.1.1(205A414246303031373730333135373133)
1-0:1.8.1(24262.769*kWh)
1-0:1.8.2(27843.037*kWh)
1-0:2.8.1(01331.436*kWh)
1-0:2.8.2(03345.015*kWh)
0-0:96.14.0(0002)
1-0:1.7.0(0000.92*kW)
1-0:2.7.0(0000.00*kW)
0-0:96.13.1()
0-0:96.13.0()
0-1:24.1.0(3)
0-1:96.1.0(4730303135353631303133373938323133)
0-1:24.3.0(211014180000)(08)(60)(1)(0-1:24.2.1)(m3)
(18533.473)
!`

**Comment by dhover** [comment]
> So this all looks exactly like what @peterforeman is looking at, except that your MBUS ID is 1 instead of 2.
So my MBUS ID = 1 where the default is 2? If so where can I set this value or is the MBUS ID value detemined during the parsing of the telegram? 

Another question: Is it possible so activate some more logging for the DSMR component so we can see where it goes wrong?

**Comment by dhover** [comment]
Now I understand the problem. Thanks for your explanation.
If I can help let me know!

**Comment by dhover** [comment]
With Slimmelezer and DSMR 2 and esphome 2021.11.1 and core-2021.11.4 the following sensors are empty:

- p1_version
- gas_consumed

**Comment by dhover** [comment]
@Skons I've tested the workaround but get an compile error: 

```
Compiling /data/slimmelezer/.pioenvs/slimmelezer/src/main.cpp.o
/config/esphome/slimmelezer.yaml: In lambda function:
/config/esphome/slimmelezer.yaml:186:62: error: 'parse_float' was not declared in this scope
            id(gas_consumed).publish_state(parse_float(s.c_str()).value());
                                                              ^
*** [/data/slimmelezer/.pioenvs/slimmelezer/src/main.cpp.o] Error 1
```

**Comment by peterforeman** [comment]
I fixed it with:

id(gas_delivery_total).publish_state(String(s.c_str()).toFloat());

**Comment by mmakaay** [comment]
I think the direct replacement for `parse_float(...)` would be `parse_number<float>(...)` from `helpers.h`.

**Comment by dhover** [comment]
@peterforeman and @mmakaay
I've tested both solutions and they both work.
Magnificent help!

**Comment by marjeroen** [comment]
can you pls paste the full working code?

**Comment by dhover** [comment]
```
text_sensor:
  - platform: dsmr
    gas_delivered_text:
      name: "Slimmelezer Gas Consumed Text"
      id: gas_consumed_text

sensor:
  - platform: dsmr
    gas_delivered:
      name: "Slimmelezer Gas Consumed"
      state_class: total_increasing
      id: gas_consumed 

  - platform: template
    name: gas_processor
    id: gas_processor
    internal: true
    update_interval: 10s
    lambda: |-
      std::string str = id(gas_consumed_text).state;
      std::vector<std::string> v;
      char * token;
      char seps[] = "()";
      token = strtok (&str[0],seps);
      while (token != NULL)
      {
        v.push_back(token);
        token = strtok (NULL, seps);
      }
      int count = 0;
      for ( std::string s : v )
      {
         if (count == 6) {
           id(gas_consumed).publish_state(parse_number<float>(s.c_str()).value());
         }
         count++;
      }
      return {""}; 
```

**Comment by peterforeman** [comment]
Unfortunately, the latest ESPhome version (2021.12.3) my "gas_delivered_text" is still empty.

```
[21:37:47][D][uart_debug:158]: <<< "0-2:24.3.0(220114210000)(00)(60)(1)(0-2:24.2.0)(m3)\r\n"
[21:37:47][D][uart_debug:158]: <<< "(08996.509)\r\n"
```

I've set gas_mbus_id to 2 to no avail.

**Comment by dhover** [comment]
> I've set gas_mbus_id to 2 to no avail.

I don't know if the component is working when the gas_mbus_id not 1.





> I do see that using this option will result in the following in `esphome/core/defines.h`:
> 
> ```c++
> #define DSMR_GAS_MBUS_ID 2
> ```
> 
> However, this define is not seen by the DSMR parser code. If I modify the DSMR library to have this:
> 
> ```c++
> #ifndef DSMR_GAS_MBUS_ID
> whoops, mbus ID is not defined!
> #define DSMR_GAS_MBUS_ID 1
> #endif
> ```
> 
> then I get a compile error `error: 'whoops' does not name a type`. So based on that, my conclusion is that the defines.h include is not applied here.

It looks like is not possible to set gas_mbus_id.
Can anyone confirm if it is possible to set the gas_mbus_id to a value other than 1?

**Comment by peterforeman** [comment]
> It looks like is not possible to set gas_mbus_id. Can anyone confirm if it is possible to set the gas_mbus_id to a value other than 1?

I've just tested this by changing "piolibs/Dsmr/src/dsmr/fields.h" and hardcoded DSMR_GAS_MBUS_ID to 2 and in my case this now works with above mentioned configuration.

So apparently the problem is indeed that the gas_mbus_id option is not working.

**Comment by heintjeput** [comment]
I have a similar issue, I also cannot get the gas usage from my DSMR2.
However I noted that above the gas_consumed is used, however I do not have this sensor available in esphome. Also in the [documentation](https://esphome.io/components/sensor/dsmr.html) only gas_delivered is mentioned. 
What I should note is that once I set gas_mbus_id: 2 I won't get my gas_delivered_text sensor anymore. So I believe the option is working at some level.

**Comment by dhover** [comment]
gas_delivered is an template sensor which is extracted from the Gas Delivered Text sensor in Esphome because the original one gas_delivered returns an empty value.

Maybe the following code is better to understand. Both sensors (the original faulty one and the calculated one) are available.

```
sensor:
  - platform: dsmr
    # original DSMR sensor
    gas_delivered:
      name: "Slimmelezer Gas Consumed"
      state_class: total_increasing

  - platform: template
    # extra sensor to calculate the missing gas value
    name: Slimmelezer Gas Consumed Calculated
    id: gas_consumed_calculated
    state_class: total_increasing
    accuracy_decimals: 3
    unit_of_measurement: "m³"

text_sensor:
  - platform: dsmr
    # original DSMR sensor
    gas_delivered_text:
      name: "Slimmelezer Gas Consumed Text"
      id: gas_consumed_text
      # if the state of "gas_delivered_text" changes a new value for the extra gas sensor is calculated
      on_value:
        then:
          - lambda: |-
              std::string str = id(gas_consumed_text).state;
              std::vector<std::string> v;
              char * token;
              char seps[] = "()";
              token = strtok (&str[0],seps);
              while (token != NULL)
              {
                v.push_back(token);
                token = strtok (NULL, seps);
              }
              int count = 0;
              for ( std::string s : v )
              {
                 if (count == 6) {
                   id(gas_consumed_calculated).publish_state(parse_number<float>(s.c_str()).value());
                 }   
                 count++;
              }
```

The code is different from the code published earlier.
In the previous code the calculation was done with an update interval of 10 sec. This example calculates a new value only when the gas_delivered_text gets a new value.

**Comment by peterforeman** [comment]
Update to this: I upgraded my dsmr ESP which made my change to fields.h undone. For other users who want to fix this while using ESPhome in docker container: exec into docker container and edit file /piolibs/Dsmr/src/dsmr/fields.h there. Then rebuild firmware and install.

I still need to fix this manually unfortunately.

**Comment by dhover** [comment]
@peterforeman Which editor do you use inside the ESPhome container?

**Comment by peterforeman** [comment]
> @peterforeman Which editor do you use inside the ESPhome container?

vim (apt update && apt install vim)

**Comment by dhover** [comment]
Ah you've installed it. I thought it was already installed. Thanks!

> I still need to fix this manually unfortunately.
Also the gas_delivered sensor for my DSMR2 is still not fixed but there is a workaround.
Does anyone know if there will be a fix for both issues?

**Comment by CyberFlexie** [comment]
Thank you for this! Works great on my end. I did have to add `device_class: gas` in the template part or else Homeassistant would complain about an Unexpected device class... 

> gas_delivered is an template sensor which is extracted from the Gas Delivered Text sensor in Esphome because the original one gas_delivered returns an empty value.
> 
> Maybe the following code is better to understand. Both sensors (the original faulty one and the calculated one) are available.
> 
> ```
> sensor:
>   - platform: dsmr
>     # original DSMR sensor
>     gas_delivered:
>       name: "Slimmelezer Gas Consumed"
>       state_class: total_increasing
> 
>   - platform: template
>     # extra sensor to calculate the missing gas value
>     name: Slimmelezer Gas Consumed Calculated
>     id: gas_consumed_calculated
>     state_class: total_increasing
>     accuracy_decimals: 3
>     unit_of_measurement: "m³"
> 
> text_sensor:
>   - platform: dsmr
>     # original DSMR sensor
>     gas_delivered_text:
>       name: "Slimmelezer Gas Consumed Text"
>       id: gas_consumed_text
>       # if the state of "gas_delivered_text" changes a new value for the extra gas sensor is calculated
>       on_value:
>         then:
>           - lambda: |-
>               std::string str = id(gas_consumed_text).state;
>               std::vector<std::string> v;
>               char * token;
>               char seps[] = "()";
>               token = strtok (&str[0],seps);
>               while (token != NULL)
>               {
>                 v.push_back(token);
>                 token = strtok (NULL, seps);
>               }
>               int count = 0;
>               for ( std::string s : v )
>               {
>                  if (count == 6) {
>                    id(gas_consumed_calculated).publish_state(parse_number<float>(s.c_str()).value());
>                  }   
>                  count++;
>               }
> ```
> 
> The code is different from the code published earlier. In the previous code the calculation was done with an update interval of 10 sec. This example calculates a new value only when the gas_delivered_text gets a new value.

**Comment by leventkomurcu** [comment]
Anyone who is having problems with gas_mbus_id set to other than 1 can try following work around. I have already submitted a PR for it.

Add following to your yaml file under esphome section. This way you don't have to modify the header file every time there is an update to esphome.

esphome:
  platformio_options:
    build_flags:
      - -DDSMR_GAS_MBUS_ID=2
